### PR TITLE
feat(Click Up): Automatically update status of clickup cards (#4prrm9)

### PR DESCRIPTION
**Pull request by VincentMolinie**
# What is Done ?
- The status will be set to `doing` when creating a pull request if that's not already the case
- The status will pass to `waiting for review` when the assignee is a reviewer
- The status will pass to `on code review` once a review has been submitted
- The status will pass to `to release` once approved
- The status will pass to `released` once push on master

# What does not work yet
- The update of subtasks
  - I'm waiting for a reply from Click Up Support

# What is required for this action to works
Simply the task id in the title of the pull request and in the title when merging too.

# How does it works with subtasks
A task with subtasks will never be attached to a PR, only the subtasks. So whenever a subtasks is updated the parent will correctly be updated depending on all the subtasks
And when a tasks pass on released state the subtasks will too

# How to test
- Create a fake  card on click up. (Let the status in `todo`)
- Create a fake pull request on top of this one and in the title put `(#${ID_OF_MY_FAKE_TASK})`
- The status of the card should be in `doing` now
- Ask a review from someone on your fake PR
- The task should pass to `waiting for review`
- Ask the reviewer to do one comment and reject for example
- The task should be `in code review`
- Then ask the reviewer to approve the PR
- The task should be on `to release`

### CODE REVIEW

File: scripts/update-clickup-status.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
+const fs = require('fs');
+const axios = require('axios');
+
+const file = fs.readFileSync(process.env.GITHUB_EVENT_PATH, 'utf8');
```
Comment (slimee) [inline scripts/update-clickup-status.js]: [naming] which file?

File: scripts/update-clickup-status.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    return eventPayload.review.state === 'APPROVED';
+  }
+  return false;
+}
+
+if (eventPayload && eventPayload.pull_request && eventPayload.pull_request.title) {
+  if (eventPayload.pull_request.title.includes('(#')) {
+    const clickUpTaskId = getClickUpTaskIdFromTitle(eventPayload.pull_request.title);
+
+    let targetStatus = 'doing';
```
Comment (slimee) [inline scripts/update-clickup-status.js]: It can't stay in 'todo' ?

File: scripts/update-clickup-status.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      if (isInCodeReview()) {
+        targetStatus = 'in code review';
+      }
+    }
+
+    if (isApproved()) {
+      targetStatus = 'to release';
+    }
+
+    axios({
```
Comment (jeffladiray) [inline scripts/update-clickup-status.js]: [Suggestion] Shouldn't this be refactored with the previous `PUT` call ?

File: scripts/update-clickup-status.js (hunk 1 above)
Comment (VincentMolinie) [reply scripts/update-clickup-status.js]: I'm going to comment the file and implement the missing hooks 😉

File: scripts/update-clickup-status.js (hunk 2 above)
Comment (VincentMolinie) [reply scripts/update-clickup-status.js]: No this mean you have started to work on it so this makes no sense to let it in todo ;)

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -0,0 +1,24 @@
+name: A workflow to update click up tasks status
```
[the remark this answers is not in the document]
Comment (VincentMolinie) [reply .github/workflows/main.yml]: I don't know. It was easy to do it like that. I absolutely don't know if it is that easy with travis

File: package.json [hunk 5]
```diff
@@ -3,7 +3,6 @@
   "version": "2.0.0",
   "description": "Forest Admin internal DevOps toolbox",
   "main": "index.js",
-  "type": "module",
```
[the remark this answers is not in the document]
Comment (VincentMolinie) [reply package.json]: Yes. The script crash if we let that line

File: scripts/update-clickup-status.js (hunk 3 above)
Comment (VincentMolinie) [reply scripts/update-clickup-status.js]: I changed it 😉

**Comment by slimee** [comment]
Which repo did you already add this feature?

**Comment by VincentMolinie** [comment]
> I've create this PR on devops repo: #10
> The name of the PR is: "Fake - Test for clickup integration (#4akzg5)"
> It is related to this clickup card: https://app.clickup.com/t/4akzg5
> 
> Sadly, the card is still in "todo". (5 minutes after PR builds end, maybe need to wait more?)

Yeah but what action are you expecting ahah. You are based on devel, the action exist only in this PR right now so of course it won't run on your fake PR. You need to rebase it on this PR

**Comment by slimee** [comment]
I give a new test after rebase (you had to explain to me that it's on PR and not on the entire repo!! 👿ah ha)
Let's have a look on this run log, it has a 400 error in `yarn run updateClickupStatus` 
https://github.com/ForestAdmin/devops/pull/10/checks?check_run_id=549104274

**Comment by VincentMolinie** [comment]
> I give a new test after rebase (you had to explain to me that it's on PR and not on the entire repo!! 👿ah ha)
> Let's have a look on this run log, it has a 400 error in `yarn run updateClickupStatus`
> https://github.com/ForestAdmin/devops/pull/10/checks?check_run_id=549104274

I fixed the issue ;)

**Pull request merged**